### DISCUSSION TITLE
Should forbidding named JSON import be done in the parser?

### DISCUSSION THREAD

**Discussion by Josh-Cena** [babel/babel#14816] [General]
https://github.com/babel/babel/pull/14668 added a parser error when attempting to do named imports from JSON modules.

I'm confident that this change is in the interest of the community and could be useful to catch errors. However, from a semantic point of view, the spec draft (https://tc39.es/proposal-json-modules/) doesn't forbid `import { foo } from "mod" assert { type: "json" }` as an early error—as such, it should not be a parser error, merely a runtime one. https://github.com/babel/babel/pull/14668#pullrequestreview-1012768829 expressed the same concern but didn't really block the PR on that ground.

I often use the Babel parser to explore quirky JS syntax, but if the parser keeps adding early diagnostics for invalid semantics and causes no AST to be generated at all, it makes testing unnecessarily hard, since static semantics are mixed with runtime ones. It also pushes other parsers that aim to align with Babel to implement the same behavior (e.g. https://github.com/typescript-eslint/typescript-eslint/pull/5377#issuecomment-1193959576) —although that's probably a weaker argument.

Is it possible that we defer the error to the generator, and only emit the error when downleveling ES modules to CJS, etc., otherwise leaving the error to the runtime?

**Comment by liuxingbaoyu** [comment]
> I often use the Babel parser to explore quirky JS syntax, but if the parser keeps adding early diagnostics for invalid semantics and causes no AST to be generated at all, it makes testing unnecessarily hard, since static semantics are mixed with runtime ones. It also pushes other parsers that aim to align with Babel to implement the same behavior (e.g. [typescript-eslint/typescript-eslint#5377 (comment)](https://github.com/typescript-eslint/typescript-eslint/pull/5377#issuecomment-1193959576)) —although that's probably a weaker argument.

For this only, the error is recoverable, you can pass the option `{ errorRecovery: true }` to make it generate ast normally.

As for the others I'm not very familiar with, let's wait for other people's opinions.

**Comment by JLHwung** [comment]
> the spec draft (https://tc39.es/proposal-json-modules/) doesn't forbid import { foo } from "mod" assert { type: "json" } as an early error—as such, it should not be a parser error

This test262 test states the opposite:

https://github.com/tc39/test262/blob/e48df0b96457b67b2d81fab16cabb230152a6c57/test/language/import/json-named-bindings.js#L7-L10

We implement the syntax error when we are bumping up the test262 conformance. Since the spec draft has not been updated since Feb 2021, I think we should move the discussion to the spec repo and I believe the champion will clarify that.

**Reply by Josh-Cena** [reply]
I see https://github.com/tc39/proposal-json-modules/issues/27. Thanks!

**Reply by nicolo-ribaudo** [reply]
@JLHwung That tests doesn't verify that it's a parse error, just that it's an error before evaluating.

From a spec point of view, you can parse the module but it throws when you then call the Link() method of the Module Record. It throws in the same phase as it throws when importing a binding which is not exported yet.

This is currently indistinguishable from parse errors, but it will be with future proposals, for example, module blocks. This code will be valid:
```js
let mod = module {
  import { foo } from "mod" assert { type: "json" }
}
```
but it will throw as soon as you do `await import(mod)`

**Reply by Josh-Cena** [reply]
@nicolo-ribaudo I'm not familiar with test262 infra, but I see `phase: parse` in the metadata. What other phases are there?

**Reply by nicolo-ribaudo** [reply]
You are right, "parse" is before "resolution" (which is when deps are resolved): https://github.com/tc39/test262/blob/e48df0b96457b67b2d81fab16cabb230152a6c57/INTERPRETING.md#negative

cc @ptomato is that test correct?

**Reply by ptomato** [reply]
I'm not an expert on that, but from the best I could determine, this error would occur during [InitializeEnvironment](https://tc39.es/ecma262/#sec-source-text-module-record-initialize-environment) of a Source Text Module Record, so resolution seems like the correct phase. I'd recommend opening an issue in test262.
